pkg/collector/corechecks/containers/containerd.go

```diff
@@ -0,0 +1,359 @@
+		return err
+	}
+
+	c.hostname, err = util.GetHostname()
+	if err != nil {
+		return err
+	}
+
+	if err = c.instance.Parse(config); err != nil {
+		return err
+	}
+	c.sub.Filters = c.instance.Filters
+	return nil
+}
+
+// Run executes the check
+func (c *ContainerdCheck) Run() error {
+	sender, err := aggregator.GetSender(c.ID())
+	defer sender.Commit()
+	if err != nil {
+		return err
+	}
+	// As we do not rely on a singleton, we ensure connectivity every check run.
+	cu, errHealth := cutil.GetContainerdUtil()
+	if errHealth != nil {
+		log.Infof("Error ensuring connectivity with Containerd daemon %v", errHealth)
+		return errHealth
+	}
+	ns := cu.Namespace()
+
+	if c.sub == nil {
+		c.sub = CreateEventSubscriber("ContainerdCheck", ns, c.instance.Filters)
+	}
+
+	if !c.sub.IsRunning() {
+		// Keep track of the health of the Containerd socket
+		c.sub.CheckEvents(cu)
+	}
+	events := c.sub.Flush(time.Now().Unix())
+	// Process events
+	computeEvents(c.hostname, events, sender, c.instance.Tags)
+
+	nk := namespaces.WithNamespace(context.Background(), ns)
+	computeMetrics(sender, nk, cu, c.instance.Tags)
+
+	return nil
+}
+
+// compute events converts Containerd events into Datadog events
+func computeEvents(hostname string, events []ContainerdEvent, sender aggregator.Sender, userTags []string) {
+	for _, e := range events {
+		split := strings.Split(e.Topic, "/")
+		if len(split) != 3 {
+			// sanity checking the event, to avoid submitting
+			log.Tracef("Event topic %s does not have the expected format", e.Topic)
+			continue
+		}
+
+		output := metrics.Event{
+			Priority:       metrics.EventPriorityNormal,
+			Host:           hostname,
+			SourceTypeName: containerdCheckName,
+			EventType:      containerdCheckName,
+			AggregationKey: fmt.Sprintf("containerd:%s", e.Topic),
+		}
+		output.Text = e.Message
+		if len(e.Extra) > 0 {
+			for k, v := range e.Extra {
+				output.Tags = append(output.Tags, fmt.Sprintf("%s:%s", k, v))
+			}
+		}
+		output.Tags = append(output.Tags, userTags...)
+		output.Ts = e.Timestamp.Unix()
+		output.Title = fmt.Sprintf("Event on %s from Containerd", split[1])
+		if split[1] == "containers" || split[1] == "tasks" {
+			// For task events, we use the container ID in order to query the Tagger's API
+			tags, err := tagger.Tag(e.ID, true)
+			if err != nil {
+				// If there is an error retrieving tags from the Tagger, we can still submit the event as is.
+				log.Errorf("Could not retrieve tags for the container %s: %v", e.ID, err)
+			}
+			output.Tags = append(output.Tags, tags...)
+		}
+		sender.Event(output)
+	}
+}
+
+func computeMetrics(sender aggregator.Sender, nk context.Context, cu cutil.ContainerdItf, userTags []string) {
+	containers, err := cu.Containers()
+	if err != nil {
+		log.Errorf(err.Error())
+		return
+	}
+
+	for _, ctn := range containers {
+		t, errTask := ctn.Task(nk, nil)
+		if errTask != nil {
+			log.Tracef("Could not retrieve metrics from task %s: %s", ctn.ID()[:12], errTask.Error())
```

Comment:
Same, debug is legit here as we ask for flares in debug mode

Reply:
This one is actually very verbose - In staging, I would have dozen of lines repeated every check run b/c init containers are Not Running, so they are listed in the output of `cu.Containers()` but will error out in the `ctn.Task(nk,nil)`.
I tried to make something more elegant, but the Status of the container is in the Task, which is nil as it errored out.

Relevant code: https://github.com/containerd/containerd/blob/master/container.go#L360
And the status: https://github.com/containerd/containerd/blob/master/task.go#L62-L64 
part of the task, so not accessible when the container is not running but was still scheduled (ex: init container)

Reply:
In the docker check, we only try to collect performance metrics if the container is running, would that be applicable here?